internal/backend/service.go

```diff
@@ -29,11 +84,114 @@ func NewDefaultService(c Persister, l log.Logger) *DefaultService {
+	}
+
+	// find the first access request matching the requested role
+	for _, ar := range accessRequests {
+		if ar.Spec.Role.TemplateName == roleName {
+			return ar, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (s *DefaultService) ListAccessRequests(ctx context.Context, key *AccessRequestKey, shouldSort bool) ([]*api.AccessRequest, error) {
+	accessRequests, err := s.k8s.ListAccessRequests(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("error getting accessrequest from k8s: %w", err)
+	}
+
+	filtered := []*api.AccessRequest{}
+	for i, ar := range accessRequests.Items {
+		if ar.Status.RequestState == api.ExpiredStatus {
+			// ignore expired request
+			continue
+		}
+		filtered = append(filtered, &accessRequests.Items[i])
+	}
+
+	if shouldSort {
+		slices.SortStableFunc(filtered, defaultAccessRequestSort)
+	}
+
+	return filtered, nil
+}
+
+func (s *DefaultService) GetGrantingAccessBinding(ctx context.Context, roleName string, namespace string, groups []string, app *unstructured.Unstructured, project *unstructured.Unstructured) (*api.AccessBinding, error) {
+	bindings, err := s.getAccessBindings(ctx, roleName, namespace)
+	if err != nil {
+		s.logger.Error(err, "error getting access bindings")
+		return nil, fmt.Errorf("error retrieving access bindings for role %s: %w", roleName, err)
+	}
+
+	if len(bindings) == 0 {
+		return nil, nil
+	}
+
+	s.logger.Debug(fmt.Sprintf("found %d bindings referencing role %s", len(bindings), roleName))
+	var grantingBinding *api.AccessBinding
+	for i, binding := range bindings {
+
+		subjects, err := binding.RenderSubjects(app, project)
+		if err != nil {
+			s.logger.Error(err, fmt.Sprintf("cannot render subjects %s:", binding.Name))
+			continue
+		}
+
+		if s.matchSubject(subjects, groups) {
+			grantingBinding = bindings[i]
+			break
+		}
+	}
+
+	return grantingBinding, nil
+}
+
+func (s *DefaultService) matchSubject(subjects, groups []string) bool {
+	for _, subject := range subjects {
+		for _, g := range groups {
+			if subject == g {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// CreateAccessRequest implements Service.
+func (s *DefaultService) CreateAccessRequest(ctx context.Context, key *AccessRequestKey, binding *api.AccessBinding) (*api.AccessRequest, error) {
+	roleName := binding.Spec.RoleTemplateRef.Name
+	ar := &api.AccessRequest{
+		ObjectMeta: v1.ObjectMeta{
+			Namespace:    key.Namespace,
+			GenerateName: s.getAccessRequestPrefix(key.Username, roleName),
+		},
+		Spec: api.AccessRequestSpec{
+			Role: api.TargetRole{
+				TemplateName: binding.Spec.RoleTemplateRef.Name,
+				Ordinal:      binding.Spec.Ordinal,
+				FriendlyName: binding.Spec.FriendlyName,
+			},
+			Subject: api.Subject{
+				Username: key.Username,
+			},
+			Application: api.TargetApplication{
+				Name:      key.ApplicationName,
+				Namespace: key.ApplicationNamespace,
+			},
+		},
+	}
+	//TODO: Set duration. Configurable by the users? Server Config?
```

Comment:
Let's start with server config as it is simpler.

Reply:
Should be implemented in another PR